### PR TITLE
pfring v0.5: implements ips mode

### PR DESCRIPTION
**Pull request by glongo**
This PR permits to use suricata and PF_RING in ips mode,
minor issues fixed in this version.

Last PR: https://github.com/inliniac/suricata/pull/1281

Configuration for using PF_RING in IPS mode:
pfring:
- interface: p4p1
  threads: 1
  cluster-id: 99
  cluster-type: cluster_flow
  copy-mode: ips
  copy-iface: eth1
  flow-packet: yes

- interface: eth1
  threads: 1
  cluster-id: 99
  cluster-type: cluster_flow
  copy-mode: ips
  copy-iface: p4p1
  flow-packet: yes

- PR build: https://buildbot.openinfosecfoundation.org/builders/glongo/builds/20
- PR pcaps: https://buildbot.openinfosecfoundation.org/builders/glongo-pcap/builds/19

### CODE REVIEW

File: src/runmode-pfring.c [hunk 1]
```diff
@@ -370,6 +373,44 @@ void *ParsePfringConfig(const char *iface)
...
 
+    if (ConfGetChildValueWithDefault(if_root, if_default, "copy-iface", &out_interface) == 1) {
+        if (strlen(out_interface) > 0) {
+            pfconf->out_interface = out_interface;
+        }
+    }
+
+    if (ConfGetChildValueWithDefault(if_root, if_default, "copy-mode", &copymodestr) == 1) {
+        if (pfconf->out_interface == NULL) {
+            SCLogInfo("Copy mode activated bu no destination"
```
Comment (inliniac) [inline src/runmode-pfring.c]: typo "bu"

File: src/runmode-pfring.c [hunk 2]
```diff
@@ -370,6 +373,44 @@ void *ParsePfringConfig(const char *iface)
...
+                    iface,
+                    pfconf->out_interface);
+            pfconf->copy_mode = PFRING_COPY_MODE_IPS;
+        } else if (strcmp(copymodestr, "tap") == 0) {
+            SCLogInfo("PF_RING TAP mode activated %s->%s",
+                    iface,
+                    pfconf->out_interface);
+            pfconf->copy_mode = PFRING_COPY_MODE_TAP;
+        } else {
+            SCLogInfo("Invalid mode (not in tap, ips)");
```
Comment (inliniac) [inline src/runmode-pfring.c]: I think all errors in this code should just that: errors. This is an issue in the afpacket code as well.

File: src/source-pfring.c [hunk 3]
```diff
@@ -209,6 +345,47 @@ static inline void PfringDumpCounters(PfringThreadVars *ptv)
...
+    }
+
+    if (p->pfring_v.peer->flags & PFRING_RING_PROTECT)
+        SCMutexLock(&p->pfring_v.peer->ring_protect);
+
+    r = pfring_send(p->pfring_v.peer->pd, (char *)pkt_buffer, buffer_size, p->pfring_v.flush_packet);
+    if (r < 0) {
+        SCLogWarning(SC_ERR_SOCKET, "Sending packet failed [%s->%s]: %s",
+                    p->pfring_v.iface, p->pfring_v.peer->iface,
+                    strerror(errno));
```
Comment (inliniac) [inline src/source-pfring.c]: Does pfring_send set errno?

File: src/source-pfring.c [hunk 4]
```diff
@@ -274,6 +451,18 @@ static inline void PfringProcessPacket(void *user, struct pfring_pkthdr *h, Pack
     }
 
     SET_PKT_LEN(p, h->caplen);
+
+    p->ReleasePacket = PfringReleasePacket;
+    p->pfring_v.copy_mode = ptv->copy_mode;
+    p->pfring_v.flush_packet = ptv->flush_packet;
+    p->pfring_v.mpeer = ptv->mpeer;
+    strlcpy(p->pfring_v.iface, ptv->interface, PFRING_IFACE_NAME_LENGTH);
```
Comment (inliniac) [inline src/source-pfring.c]: Why is this necessary?

File: src/runmode-pfring.c (hunk 1 above)
Comment (glongo) [inline src/runmode-pfring.c]: What do you think to add an error code here?

File: src/runmode-pfring.c (hunk 2 above)
Comment (glongo) [inline src/runmode-pfring.c]: ack

File: src/source-pfring.c (hunk 3 above)
Comment (glongo) [inline src/source-pfring.c]: yes it does

File: src/runmode-pfring.c (hunk 1 above)
Comment (inliniac) [inline src/runmode-pfring.c]: Yeah please do. AF_PACKET has the same issue, and I recently spent too much time debugging why it did not forward packets.

File: src/source-pfring.c (hunk 4 above)
Comment (glongo) [inline src/source-pfring.c]: It's not really necessary, see line 365, just to print the interfaces

Comment (inliniac) [inline src/source-pfring.c]: I think you should be able to get this from p->pfring_v.mpeer?

Comment (glongo) [inline src/source-pfring.c]: Yes, I've tried now.
I can remove it
